File: examples/persist-session/rest/nodejs/persist-session.mjs
// Creates a long-lived Browserless session, connects to set state, disconnects,
// then reconnects to verify the state persists.
//
// Install: npm install puppeteer-core
// Run:     node persist-session.mjs

import puppeteer from "puppeteer-core";

const TOKEN = "YOUR_API_TOKEN_HERE";

// Step 1: Create a long-lived session
const sessionResponse = await fetch(
  `https://production-sfo.browserless.io/session?token=${TOKEN}`,
  {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ttl: 300000, stealth: true, headless: false }),
  }
);
const session = await sessionResponse.json();
console.log("Session created:", session.id);

// Step 2: Connect via Puppeteer and store data
const browser = await puppeteer.connect({ browserWSEndpoint: session.connect });
const page = await browser.newPage();

await page.goto("https://automationexercise.com", { waitUntil: "networkidle2" });

await page.evaluate(() => {
  localStorage.setItem(
    "shoppingCart",
    JSON.stringify({ items: [{ id: 1 }], totalItems: 1, totalPrice: 500 })
  );
  localStorage.setItem("userPreferences", JSON.stringify({ theme: "dark" }));
});

await browser.disconnect();
console.log("State set. Disconnected.");

// Step 3: Reconnect and verify state persists
const browser2 = await puppeteer.connect({ browserWSEndpoint: session.connect });
const page2 = await browser2.newPage();
await page2.goto("https://automationexercise.com", { waitUntil: "networkidle2" });

const cart = await page2.evaluate(() => localStorage.getItem("shoppingCart"));
console.log("Cart persisted:", JSON.parse(cart));

await browser2.close();

// Step 4: Delete the session
await fetch(`${session.stop}&force=true`, { method: "DELETE" });
console.log("Session stopped.");

File: examples/persist-session/rest/python/persist_session.py
# Creates a long-lived Browserless session, connects to set state, disconnects,
# then reconnects to verify the state persists.
#
# Install: pip install requests playwright
# Run:     python persist_session.py

import requests
from playwright.sync_api import sync_playwright

TOKEN = "YOUR_API_TOKEN_HERE"

# Step 1: Create a long-lived session
session_response = requests.post(
    f"https://production-sfo.browserless.io/session?token={TOKEN}",
    json={"ttl": 300000, "stealth": True, "headless": False},
)
session = session_response.json()
print("Session created:", session["id"])
print("Connect URL:", session["connect"])

# Step 2: Connect and set state via Playwright
with sync_playwright() as p:
    browser = p.chromium.connect_over_cdp(session["connect"])
    context = browser.contexts[0] if browser.contexts else browser.new_context()
    page = context.new_page()
    page.goto("https://automationexercise.com")

    page.evaluate("""() => {
        localStorage.setItem('shoppingCart', JSON.stringify({items: [{id: 1}]}));
    }""")

    browser.close()

    # Step 3: Reconnect and verify
    browser2 = p.chromium.connect_over_cdp(session["connect"])
    context2 = browser2.contexts[0] if browser2.contexts else browser2.new_context()
    page2 = context2.new_page()
    page2.goto("https://automationexercise.com")

    cart = page2.evaluate("() => localStorage.getItem('shoppingCart')")
    print("Cart persisted:", cart)

    browser2.close()

# Step 4: Stop the session
requests.delete(f"{session['stop']}&force=true")
print("Session stopped.")
